File: ui-fe/src/api_client.py
"""API client for communicating with Lightship MVP backend server."""
import requests
import json
import os
from typing import Dict, Any, Optional
import logging

logger = logging.getLogger(__name__)


class APIClient:
    """Client for Lightship MVP API."""

    def __init__(self, base_url: str = None):
        """Initialize API client.

        Args:
            base_url: Base URL of the API server. Falls back to BACKEND_API_URL env var or localhost.
        """
        if base_url is None:
            base_url = os.environ.get("BACKEND_API_URL", "http://localhost:8000")
        self.base_url = base_url.rstrip("/")
        self.session = requests.Session()

    def check_health(self) -> bool:
        """Check if API server is healthy.

        Returns:
            True if server is reachable and healthy
        """
        try:
            # Lambda cold start can take 60-90s (model loading), use generous timeout
            response = self.session.get(f"{self.base_url}/health", timeout=120)
            return response.status_code == 200
        except Exception as e:
            logger.error(f"Health check failed: {e}")
            return False

    def upload_video(
        self,
        video_file,
        config: Optional[Dict[str, Any]] = None
    ) -> Optional[str]:
        """Upload video for processing.

        Args:
            video_file: File-like object or UploadedFile
            config: Processing configuration

        Returns:
            job_id if successful, None otherwise
        """
        try:
            # Reset file pointer
            video_file.seek(0)

            files = {
                'video': (video_file.name, video_file, 'video/mp4')
            }

            data = {}
            if config:
                data['config'] = json.dumps(config)

            response = self.session.post(
                f"{self.base_url}/process-video",
                files=files,
                data=data,
                timeout=30
            )

            if response.status_code == 200:
                result = response.json()
                return result.get('job_id')
            else:
                logger.error(f"Upload failed: {response.status_code} - {response.text}")
                return None

        except Exception as e:
            logger.error(f"Upload error: {e}")
            return None

    def get_status(self, job_id: str) -> Optional[Dict[str, Any]]:
        """Get processing status for a job.

        Args:
            job_id: Job identifier

        Returns:
            Status dictionary or None if failed
        """
        try:
            response = self.session.get(
                f"{self.base_url}/status/{job_id}",
                timeout=10
            )

            if response.status_code == 200:
                return response.json()
            else:
                logger.error(f"Get status failed: {response.status_code}")
                return None

        except Exception as e:
            logger.error(f"Get status error: {e}")
            return None

    def get_results(self, job_id: str) -> Optional[Dict[str, Any]]:
        """Get processing results for a completed job.

        Args:
            job_id: Job identifier

        Returns:
            Results dictionary or None if failed
        """
        try:
            response = self.session.get(
                f"{self.base_url}/results/{job_id}",
                timeout=10
            )

            if response.status_code == 200:
                return response.json()
            else:
                logger.error(f"Get results failed: {response.status_code}")
                return None

        except Exception as e:
            logger.error(f"Get results error: {e}")
            return None

    def get_json_content(self, job_id: str) -> Optional[Dict[str, Any]]:
        """Get output JSON content.

        Args:
            job_id: Job identifier

        Returns:
            JSON data or None if failed
        """
        try:
            response = self.session.get(
                f"{self.base_url}/download/json/{job_id}",
                timeout=10
            )

            if response.status_code == 200:
                return response.json()
            else:
                logger.error(f"Get JSON failed: {response.status_code}")
                return None

        except Exception as e:
            logger.error(f"Get JSON error: {e}")
            return None

    def download_frame(self, job_id: str, frame_idx: int, save_path: str) -> bool:
        """Download frame image.

        Args:
            job_id: Job identifier
            frame_idx: Frame index
            save_path: Path to save image

        Returns:
            True if successful
        """
        try:
            response = self.session.get(
                f"{self.base_url}/download/frame/{job_id}/{frame_idx}",
                timeout=10
            )

            if response.status_code == 200:
                with open(save_path, 'wb') as f:
                    f.write(response.content)
                return True
            else:
                logger.error(f"Download frame failed: {response.status_code}")
                return False

        except Exception as e:
            logger.error(f"Download frame error: {e}")
            return False

    def list_jobs(self, limit: int = 50) -> list:
        """List recent jobs from the backend.

        Args:
            limit: Maximum number of jobs to return

        Returns:
            List of job dicts, newest first. Empty list on failure.
        """
        try:
            response = self.session.get(
                f"{self.base_url}/jobs",
                params={"limit": limit},
                timeout=10
            )
            if response.status_code == 200:
                return response.json().get("jobs", [])
            else:
                logger.error(f"List jobs failed: {response.status_code}")
                return []
        except Exception as e:
            logger.error(f"List jobs error: {e}")
            return []

    def cleanup(self, job_id: str) -> bool:
        """Cleanup temporary files for a job.

        Args:
            job_id: Job identifier

        Returns:
            True if successful
        """
        try:
            response = self.session.delete(
                f"{self.base_url}/cleanup/{job_id}",
                timeout=10
            )

            return response.status_code == 200

        except Exception as e:
            logger.error(f"Cleanup error: {e}")
            return False

File: ui-fe/src/lib/api-client.ts
/**
 * Backend API client for the Lightship pipeline.
 *
 * Replaces mock-results.ts with real HTTP calls to the FastAPI backend.
 * The backend serves PipelineResultJson at GET /pipeline-result/{jobId}.
 */

import type {
  PipelineResultJson,
  PipelineConfig,
  AssetResult,
  ResultPropertyRow,
} from '@/components/evaluation/flow-types';

const API_BASE =
  typeof window !== 'undefined'
    ? (process.env.NEXT_PUBLIC_API_BASE ?? '')
    : '';

async function apiFetch<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${API_BASE}${path}`, {
    ...init,
    headers: {
      'Content-Type': 'application/json',
      ...init?.headers,
    },
  });

  if (!res.ok) {
    throw new Error(`API ${res.status}: ${res.statusText}`);
  }

  return res.json() as Promise<T>;
}

export async function healthCheck(): Promise<boolean> {
  try {
    await apiFetch('/health');
    return true;
  } catch {
    return false;
  }
}

export type PresignResponse = {
  upload_url: string;
  s3_key: string;
};

export async function getPresignedUploadUrl(
  filename: string,
): Promise<PresignResponse> {
  return apiFetch<PresignResponse>(
    `/presign-upload?filename=${encodeURIComponent(filename)}`,
  );
}

export async function uploadToPresignedUrl(
  url: string,
  file: File,
): Promise<void> {
  const res = await fetch(url, {
    method: 'PUT',
    body: file,
    headers: {
      'Content-Type': file.type || 'video/mp4',
    },
  });

  if (!res.ok) {
    throw new Error(`S3 upload failed: ${res.status}`);
  }
}

export type ProcessVideoRequest = {
  s3_key: string;
  filename: string;
  config?: {
    snapshot_strategy?: string;
    max_snapshots?: number;
    s3_output_path?: string;
  };
};

export type ProcessVideoResponse = {
  job_id: string;
  status: string;
  message: string;
};

export async function processVideo(
  req: ProcessVideoRequest,
): Promise<ProcessVideoResponse> {
  return apiFetch<ProcessVideoResponse>('/process-video', {
    method: 'POST',
    body: JSON.stringify(req),
  });
}

export type JobStatus = {
  status: string;
  progress: number;
  current_stage?: string;
  stages_completed?: number;
  total_stages?: number;
  message?: string;
};

export async function getJobStatus(jobId: string): Promise<JobStatus> {
  return apiFetch<JobStatus>(`/status/${jobId}`);
}

export async function getPipelineResult(
  jobId: string,
): Promise<PipelineResultJson> {
  return apiFetch<PipelineResultJson>(`/pipeline-result/${jobId}`);
}

export type HistoricalJob = {
  job_id: string;
  status: string;
  filename?: string;
  video_class?: string;
  road_type?: string;
  created_at?: string;
  completed_at?: string;
  s3_results_uri?: string;
  current_stage?: string;
  progress?: string;
};

export async function listJobs(
  limit: number = 50,
): Promise<Array<HistoricalJob>> {
  return apiFetch<Array<HistoricalJob>>(`/jobs?limit=${limit}`);
}

export async function downloadJson(jobId: string): Promise<unknown> {
  return apiFetch(`/download/json/${jobId}`);
}

/**
 * Upload a file and start processing. Returns the job_id.
 */
export async function uploadAndProcess(
  file: File,
  config: PipelineConfig,
): Promise<string> {
  let s3Key: string;

  try {
    const presign = await getPresignedUploadUrl(file.name);
    await uploadToPresignedUrl(presign.upload_url, file);
    s3Key = presign.s3_key;
  } catch {
    const formData = new FormData();
    formData.append('file', file);
    formData.append(
      'config',
      JSON.stringify({
        snapshot_strategy:
          config.frameSelectionMethod === 'scene-change'
            ? 'scene_change'
            : config.frameSelectionMethod === 'native'
              ? 'naive'
              : 'clustering',
        max_snapshots: parseInt(config.nativeFps, 10) || 5,
        s3_output_path: config.s3BucketPath,
      }),
    );

    const res = await fetch(`${API_BASE}/process-video`, {
      method: 'POST',
      body: formData,
    });

    if (!res.ok) {
      throw new Error(`Upload failed: ${res.status}`);
    }

    const data = await res.json();
    return data.job_id;
  }

  const resp = await processVideo({
    s3_key: s3Key,
    filename: file.name,
    config: {
      snapshot_strategy:
        config.frameSelectionMethod === 'scene-change'
          ? 'scene_change'
          : config.frameSelectionMethod === 'native'
            ? 'naive'
            : 'clustering',
      max_snapshots: parseInt(config.nativeFps, 10) || 5,
      s3_output_path: config.s3BucketPath,
    },
  });

  return resp.job_id;
}

/**
 * Convert PipelineResultJson to the AssetResult shape used by the UI.
 */
export function toAssetResult(
  assetId: string,
  assetName: string,
  previewUrl: string,
  kind: 'video' | 'image',
  raw: PipelineResultJson,
): AssetResult {
  const rows: Array<ResultPropertyRow> = [
    { label: 'Filename', value: raw.filename },
    { label: 'Video Class', value: raw.video_class },
    { label: 'Road Type', value: raw.road_type },
    { label: 'Weather', value: raw.weather },
    { label: 'Traffic', value: raw.traffic },
    { label: 'Speed', value: raw.speed },
    { label: 'Description', value: raw.video_description || '—' },
    { label: 'Frames Analyzed', value: String(raw.frames.length) },
    {
      label: 'Total Objects',
      value: String(raw.frames.reduce((sum, f) => sum + f.objects.length, 0)),
    },
    { label: 'Hazards', value: String(raw.hazards.length) },
  ];

  return {
    assetId,
    assetName,
    previewUrl,
    kind,
    rawJson: raw,
    propertyRows: rows,
  };
}
